src/metal/window-proxy.ts

```diff
@@ -43,19 +47,35 @@ let W: WindowProxy = {
   getScrollLeft: nop,
   getHeight: nop,
   getWidth: nop,
+  rootsList: [],
   onWindowIsDirtyListeners: [],
   rAF: hasRAF ? window.requestAnimationFrame.bind(window) : (callback: Function) => { callback(); },
+  bindRootScrollEvent: (root: Element) => { root.addEventListener('scroll', () => W.callIsDirtyListeners(), false); },
+  callIsDirtyListeners() {
+    if (this.onWindowIsDirtyListeners.length <= 0) { return; }
+
+    this.onWindowIsDirtyListeners.forEach((obj: any) => {
+      let { fn, scope } = obj;
```

Comment:
rather then `scope` lets just have a bound `fn`. That way we don't need to store scope nor rebind `fn`, and if we use a map for `onWindowIsDirtyListeners` we would also no longer require `id`.

Reply:
Actually, wouldn't a simpler implementation be a single global version?

Specifically:

* globally (within one of the spaniel modules) we would store: `let VERSION = 0`
* `VERSION` would be incremented when a dirty handler is invoked (scroll/resize etc)
* on each Scheduler `ElementScheduler` we would store `this._lastVersion` and have an accessor of `get isDirty() { return VERSION !== this._lastVersion; }`

Then we would no longer need to register any `onWindowIsDirtyListeners` at all, which is a nice simplification as no book keeping mitigates many issues

----

```js
class  ElementSchedular {
  protected _lastVersion = VERSION;
  // ... snip ...  
  get isDirty() {
     return VERSION !== this._lastVersion; 
  }
  applyQueue(frame: Frame) {
    if (!this.isDirty) { return; } // also we should likely put this dirty guard before we do any work, rather then mid for loop bellow.
    this._lastVersion = VERSION;

    for (let i = 0; i < this.queue.items.length; i++) {
      let { callback, el, id, clientRect } = this.queue.items[i];

      if ( !clientRect) {
        clientRect = this.queue.items[i].clientRect = getBoundingClientRect(el);
      }

      callback(frame, id, clientRect);
    }

  }

}
```